### PR TITLE
operator: print callstack when exiting before preparing done

### LINKED ISSUE #51370: disk_snapshot_backup: print more debug info when the init pod get interrupted

### PR DESCRIPTION
**Pull request by YuJuncen**
### What problem does this PR solve?

Issue Number: close #51370 

Problem Summary:
It is hard to debug what happens when the init pod get timed out.

### What changed and how does it work?
This PR will print all gorouintes to `stdout` once the `prepare-for-snapshot-backup` command exits before finishing the full prepare stage.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

When we are not yet finished:
CleanShot 2024-02-27 at 19 28 14@2x
When we have finished prepare stage:
CleanShot 2024-02-27 at 19 28 51@2x

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/utils/misc.go [hunk 1]
```diff
@@ -157,3 +165,52 @@ func WithCleanUp(errOut *error, timeout time.Duration, fn func(context.Context)
...
+}
+
+var (
+	DumpGoroutineWhenExit atomic.Bool
+)
+
+func StartExitSingleListener(ctx context.Context) (context.Context, context.CancelFunc) {
+	cx, cancel := context.WithCancel(ctx)
+	sc := make(chan os.Signal, 1)
+	signal.Notify(sc,
```
Comment (BornChanger) [inline br/pkg/utils/misc.go:187]: To confirm, will a normal init job exit print the go routines also?

File: br/pkg/utils/misc.go [hunk 2]
```diff
@@ -157,3 +165,52 @@ func WithCleanUp(errOut *error, timeout time.Duration, fn func(context.Context)
...
+		sig := <-sc
+		dumpGoroutine := DumpGoroutineWhenExit.Load()
+		padding := strings.Repeat("=", 8)
+		printDelimate := func(s string) {
+			fmt.Printf("%s[ %s ]%s\n", padding, s, padding)
+		}
+		fmt.Println()
+		printDelimate(fmt.Sprintf("Got signal %v to exit.", sig))
+		printDelimate(fmt.Sprintf("Required Goroutine Dump = %v", dumpGoroutine))
+		if DumpGoroutineWhenExit.Load() {
```
Comment (WangLe1321) [inline br/pkg/utils/misc.go:202]: Why not just use `dumpGoroutine`?

File: br/pkg/utils/misc.go [hunk 3]
```diff
@@ -157,3 +165,52 @@ func WithCleanUp(errOut *error, timeout time.Duration, fn func(context.Context)
...
+			printDelimate("Start Dumping Goroutine")
+			os.Stdout.Write(AllStackInfo())
```
Comment (WangLe1321) [inline br/pkg/utils/misc.go:204]: Why not use `fmt.Println(AllStackInfo())`

File: br/pkg/utils/misc.go (hunk 1 above)
Comment (YuJuncen) [reply br/pkg/utils/misc.go:187]: No, goroutine dump won't be printed once prepare done.

File: br/pkg/utils/misc.go (hunk 3 above)
Comment (YuJuncen) [reply br/pkg/utils/misc.go:204]: `debug.PrintStack()` uses `write`. I think there isn't many difference between them.

**Review by WangLe1321 — APPROVED** [review]

**Review by BornChanger — APPROVED** [review]

File: br/pkg/task/operator/cmd.go [hunk 4]
```diff
@@ -141,6 +142,7 @@ func AdaptEnvForSnapshotBackup(ctx context.Context, cfg *PauseGcConfig) error {
 		if cfg.OnAllReady != nil {
 			cfg.OnAllReady()
 		}
+		utils.DumpGoroutineWhenExit.Store(false)
```
Comment (3pointer) [inline br/pkg/task/operator/cmd.go:145]: it seems no need set it back

**Review by 3pointer — COMMENTED** [review]
rest LGTM

**Pull request merged**